design/healcheck-endpoint.md

```diff
@@ -0,0 +1,159 @@
+# Healthcheck endpoint
+
+* **Status**: Draft #1
+* **JIRA**: [KEYCLOAK-12398](https://issues.redhat.com/browse/KEYCLOAK-12398)
+* **Observability JIRA**: [KEYCLOAK-8288](https://issues.jboss.org/browse/KEYCLOAK-8288)
+
+## Motivation
+
+Deploying both Keycloak and RHSSO on top of Kubernetes requires defining both Readiness Check and Liveness Check
+(see the [manual](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)).
+Currently, for Keycloak we use a very naive implementation, that calls `/auth/realms/master` URL.
+This approach has many drawbacks, here are a few examples:
+
+* We cannot discover database failures
+* We cannot discover if certain Identity Provider is up and running
+* Undertow subsystem is down when proceeding with migration
+
+In order to address these problems, we need provide a parametrized Healthcheck endpoint in the Server and use it in our
+container image as well as in the Operator.
+
+## Parameterizing the Healthcheck URL
+
+The new Health Endpoint need to support query parameters in the endpoint in the following form:
+
+    http://localhost:8080/auth/health?timeout=2s&fail-on=db,idp-google
```

Comment:
Is an example of a URL to be set for Kubernetes Readiness/Liveness probe?

Is it based on having a "generic" health check to feed the parameters into for execution? Is there concern it leaves the URL prone to breakage from incorrect query parameters?

Reply:
> Is an example of a URL to be set for Kubernetes Readiness/Liveness probe?

Yes.

> Is it based on having a "generic" health check to feed the parameters into for execution? Is there concern it leaves the URL prone to breakage from incorrect query parameters?

That's an interesting angle. I believe we'll expose our own Servlet/JAX-RS Endpoint to expose the Healthcheck. In such case, we could emit HTTP 4xx to indicate user error. 

As for the Kubernetes deployments - our suggested way of doing this is the Keycloak Operator. We will probably have some test to ensure the parameters are fine in that case.

Reply:
Let me resolve this conversation. I believe this will be covered in the second draft I plan to sent in a minute.